diff --git a/agent/session/datachannel/datachannel.go b/agent/session/datachannel/datachannel.go
index 927a040e..166111d8 100644
--- a/agent/session/datachannel/datachannel.go
+++ b/agent/session/datachannel/datachannel.go
@@ -183,6 +183,7 @@ func (dc *dataChannel) PerformHandshake(log logger.T,
 	//@ unfold dc.Mem()
 	//@ unfold dc.MemInternal(HandshakeCompleted)
 	logInfo(log, "Handshake successfully completed.")
+	log.Info(dc)
 
 	//@ unfold acc(dc.MemChannelState(), 1/2)
 	dc.dataChannelState = IODistributed
