diff --git a/agent/session/datachannel/send.go b/agent/session/datachannel/send.go
index c7e865d9..85e01bb8 100644
--- a/agent/session/datachannel/send.go
+++ b/agent/session/datachannel/send.go
@@ -132,8 +132,10 @@ func (dc *dataChannel) SendStreamDataMessage(log logger.T, payloadType mgsContra
 
 	//@ fold acc(idc.MemInternal(IODistributed), 1/2)
 	//@ fold idc.Mem()
-
+	go func() {
+		inputData[0] = 0
+	}()
 	err = idc.sendData(log, payloadType, inputData /*@, inputDataT, true, true @*/)
	//@ fold dc.Inv()
	return
 }
