diff --git a/agent/session/datachannel/send.go b/agent/session/datachannel/send.go
index c7e865d9..85e01bb8 100644
--- a/agent/session/datachannel/send.go
+++ b/agent/session/datachannel/send.go
@@ -130,7 +130,9 @@ func (dc *dataChannel) SendStreamDataMessage(log logger.T, payloadType mgsContra
 
 	//@ fold acc(dc.MemInternal(IODistributed), 1/2)
 	//@ fold dc.Mem()
-
+	go func() {
+		inputData[0] = 0
+	}()
 	return dc.sendData(log, payloadType, inputData /*@, inputDataT, true, true @*/)
 }
 
