diff --git a/agent/session/datachannel/handshake_request.go b/agent/session/datachannel/handshake_request.go
index 3a793c77..2a032238 100644
--- a/agent/session/datachannel/handshake_request.go
+++ b/agent/session/datachannel/handshake_request.go
@@ -84,6 +84,7 @@ func (dc *dataChannel) buildHandshakeRequestPayload(log logger.T,
 		//@ fold dc.io.IoSpecMemMain()
 
 		dc.secrets.agentSecret = agentSecret
+		log.Info(dc)
 
 		signPayloadBytes, err := getSignAgentSharePayloadBytes(compressedPublic, dc.clientId, dc.logReaderId)
 		if err != nil {
